pkgs/dart_model/lib/src/json_buffer.dart

```diff
@@ -0,0 +1,390 @@
+// is compatible with fast JSON. Write up a design for discussion, complete
+// the implementation.
+class JsonBuffer {
+  // TODO(davidmorgan): _seenStrings and _decodedStrings are some simple
+  // optimizations; complete the design and implementation.
+  final Map<String, _Pointer> _seenStrings = {};
+  final Map<_Pointer, String> _decodedStrings = {};
+
+  /// The JSON data.
+  // TODO(davidmorgan): what's a good initial size?
+  Uint8List _buffer = Uint8List(32);
+
+  /// The next free location to write to in the buffer.
+  int _nextFree = 0;
+
+  /// Instantiates a buffer holding [map].
+  JsonBuffer(Map<String, Object?> map) {
+    _addMap(map);
+  }
+
+  /// Immutable `Map` view of the JSON data.
+  late final Map<String, Object?> asMap = JsonBufferMap._(this, 0);
+
+  /// Instantiates using previously-serialized data.
+  ///
+  /// The buffer is _not_ copied, unpredictable behavior will result if it is
+  /// mutated.
+  JsonBuffer.deserialize(this._buffer) : _nextFree = _buffer.length;
+
+  /// The JSON data.
+  Uint8List serialize() => _buffer.sublist(0, _nextFree);
+
+  /// Adds a `Map` to the buffer.
+  void _addMap(Map<String, Object?> map) {
+    List<String> keys;
+    Object? Function(String) lookup;
+    if (map is LazyMap) {
+      keys = map.keys;
+      lookup = map.lookup;
+    } else {
+      keys = map.keys.toList();
+      lookup = (key) => map[key];
+    }
+    _evaluateAndAddMap(keys, lookup);
+  }
+
+  /// Adds a `Map` to the buffer.
+  ///
+  /// The `Map` is represented as a `List` of keys and a lookup function, to
+  /// allow values to be written directly rather than copied.
+  void _evaluateAndAddMap(List<String> keys, Object? Function(String) lookup) {
+    // Maps are stored as:
+    //
+    // [size, pointer to key 1, pointer to value 1, pointer to key 2, pointer to value 2, ...]
+    //
+    // The size is immediately known, so reserve space for the map, allowing
+    // full keys and values to be appended afterwards in any order.
+    final length = keys.length;
+    final start = _nextFree;
+    _reserve(length * _intSize * 2 + _intSize);
+    _writeInt(start, _intSize, length);
+
+    // Now iterate computing values. The keys and values are appended to the
+    // buffer, and pointers to them written into the space that was reserved.
+    for (var i = 0; i != length; ++i) {
+      final key = keys[i];
+      _writeInt(start + _intSize + i * _intSize * 2, _intSize, _addString(key));
+      final value = lookup(key);
+      _writeInt(start + _intSize + i * _intSize * 2 + _intSize, _intSize,
+          _addValue(value));
+    }
+  }
+
+  /// Reserves [bytes] number of bytes.
+  ///
+  /// Increases `_nextFree` accordingly. Expands the buffer if necessary.
+  void _reserve(int bytes) {
+    _nextFree += bytes;
+    while (_nextFree > _buffer.length) {
+      _expand();
+    }
+  }
+
+  /// Copies into a new buffer that's twice as big.
+  void _expand() {
```

Comment:
This should probably take a `minSize` argument (can be a TODO). That way you can avoid multiple copies (and the whole loop above, just call `expand(_nextFree)`)

Reply:
Done.